### PR TITLE
Add SlaClip adaptive clipping optimizer for DP-SGD (algorithm of ICML 2026 spotlight paper)

### PR DESCRIPTION
**Pull request by ZsyRock**
## Summary

This PR adds SlaClip as a self-contained research prototype under
`research/slaclip`.

SlaClip uses the norm budget left unused by clipping to jointly release a
private Slack Indicator—a noisy, binned estimate of the gradient-norm CDF—
together with the standard DP-SGD gradient.

The extended per-sample vector preserves the original L2 norm bound. Under the
same sampling rule, normalization, clipping threshold, and noise multiplier,
the gradient and Slack Indicator are therefore released as one Gaussian
mechanism with the same per-step privacy-accounting parameters as vanilla
DP-SGD.

## Usage

The implementation exposes two composable steps:

1. `SlaClipDPOptimizer` jointly releases the DP gradient and private Slack
   Indicator. Without a controller, the clipping threshold remains fixed.
2. `SlaClipController` consumes the released indicator and applies equations
   (28)-(30) from the paper to adapt the clipping threshold as post-processing.

Using both components implements the paper's SlaClip method. Users may also use
the first component independently to obtain private CDF information for other
DP-safe post-processing methods.

The recommended entry point is `SlaClipPrivacyEngine`, a research-scoped
subclass of Opacus `PrivacyEngine`. Users can call its inherited
`make_private()` or `make_private_with_epsilon()` APIs directly; model
wrapping, data-loader preparation, secure RNG selection, optimizer
construction, and privacy-accountant hooks are handled automatically. Full integration instructions, examples,
privacy assumptions, parameters, automatic K selection, and limitations are
provided in `research/slaclip/README.md`.

## Scope

This PR:

- is self-contained under `research/slaclip`;
- does not modify `PrivacyEngine` or other Opacus core APIs;
- supports the standard non-distributed `DPOptimizer` path;
- does not add benchmark or end-to-end training scripts;
- is intended as an experimental research prototype.

## Tests

```bash
python -m pytest -q research/slaclip
```

Result:

```text
17 passed, 12 subtests passed
```

The tests cover:
- the `SlaClipPrivacyEngine` entry point;
- inherited `make_private_with_epsilon()` support;
- automatic accountant-hook attachment;
- explicit rejection of unsupported clipping modes;
- automatic K selection from equation (36);
- equations (7)-(8) and the extended-gradient norm bound;
- agreement of the first d coordinates with native DPOptimizer;
- indicator-only operation;
- the paper clipping controller and threshold bounds;
- empty Poisson batches;
- a single accountant event for the joint release.

Black, flake8, isort, and git diff --check also pass.

## Reference

This implementation is based on:

SlaClip: Gradient Norm Slacks can be Indicator for Adaptive Clipping in DP-SGD, ICML 2026 Spotlight.

### CODE REVIEW

**Comment by ZsyRock** [comment]
Hi maintainers, thanks for taking a look. It seems the workflows are still awaiting approval and the PR has not been imported yet. Please let me know if there is anything I should do on my side.

**Comment by iden-kalemaj** [comment]
Hi @ZsyRock, for emerging research related to DP-SGD we recommend committing changes to the [research](https://github.com/meta-pytorch/opacus/tree/main/research) folder. Once a method becomes established enough in the community, we can move it into the main folder. This prevents the core Opacus code from becoming difficult to maintain with the addition of new methods.

**Comment by ZsyRock** [comment]
Thanks for the guidance. I revised the PR to move SlaClip into `research/slaclip` and removed the changes to the core `PrivacyEngine` and optimizer selection paths. The updated version is self-contained under the research folder, includes a README plus lightweight tests, and no longer modifies Opacus core APIs.

I verified the research tests locally with:

```bash
conda run -n opacus python -m pytest research/slaclip -q
```

which passes with `5 passed`.

**Comment by iden-kalemaj** [comment]
Thank you for the contribution @ZsyRock, please consider also adding a module extending PrivacyEngine that people can use as an entrypoint for accessing your method, see example [here](https://github.com/meta-pytorch/opacus/blob/main/research/disk_optimizer/KFprivacy_engine.py). I left a few other comments on the code as well.

File: research/slaclip/README.md
Comment (iden-kalemaj) [inline research/slaclip/README.md:1]: Please add some more detail about what the methodology does, and how users can use your methodology, see an example [here](https://github.com/meta-pytorch/opacus/blob/main/research/disk_optimizer/ReadMe.md).

File: research/README.md [hunk 1]
```diff
@@ -6,6 +6,8 @@ This directory contains novel methods built on top of Opacus that enhance DP-SGD
 ## Contributions
 We warmly welcome and encourage contributions of new methods! To contribute, please follow these steps:
 
+- Experimental SlaClip research prototype: [research/slaclip](slaclip)
```
Comment (iden-kalemaj) [inline research/README.md]: seems added by mistake?

**Review by iden-kalemaj — CHANGES_REQUESTED** [review]

File: research/slaclip/README.md
Comment (ZsyRock) [reply research/slaclip/README.md:1]: Thanks for the feedback. I substantially expanded
`research/slaclip/README.md`.

It now explains the methodology and privacy argument, and documents the two
composable usage modes: releasing only the private Slack Indicator as CDF
information, and using that indicator with the paper's controller to adapt the
clipping threshold.

I also added integration instructions, automatic K selection, parameters,
limitations, tests, and citation information.

File: research/README.md (hunk 1 above)
Comment (ZsyRock) [reply research/README.md]: Thanks, this was added by mistake and has now been removed. The final PR does
not modify `research/README.md`.

**Comment by ZsyRock** [comment]
Thanks for the review. I have updated the implementation and documentation
based on the feedback.

The PR is now a self-contained research prototype under `research/slaclip`
with no Opacus core changes. The README includes the methodology, privacy
argument, two composable usage modes, PrivacyEngine integration instructions,
automatic K selection, parameters, limitations, tests, and citation.

The branch has also been rebased onto the latest `main`, with one commit and
four changed files. The GitHub Actions workflows currently show
`action_required`; it appears they need approval before the jobs can start.

Could you please re-review the updated PR and approve the workflow runs when
convenient?

**Comment by ZsyRock** [comment]
> Hi @ZsyRock, for emerging research related to DP-SGD we recommend committing changes to the [research](https://github.com/meta-pytorch/opacus/tree/main/research) folder. Once a method becomes established enough in the community, we can move it into the main folder. This prevents the core Opacus code from becoming difficult to maintain with the addition of new methods.

Thanks for the suggestion. I added a research-scoped
`SlaClipPrivacyEngine` entry point in commit 170ccdd.

Users can now call the standard `make_private()` or
`make_private_with_epsilon()` APIs directly. The entry point automatically
constructs `SlaClipDPOptimizer` and preserves Opacus model wrapping,
data-loader preparation, secure RNG selection, and privacy-accountant hook
attachment.

Omitting the controller enables the indicator-only mode, while passing
`SlaClipController` enables the complete paper SlaClip method. I also updated
the README with both usage modes and added integration tests.

The implementation remains entirely under `research/slaclip` and does not
modify Opacus core APIs.

**Comment by ZsyRock** [comment]
Hi @iden-kalemaj, just a gentle follow-up on PR #824.  
I addressed the previous review feedback in the current head 170ccdd: the implementation is now fully scoped under research/slaclip, the README has been expanded, and a research-scoped SlaClipPrivacyEngine entry point has been added. Both review threads are resolved, and the branch is up to date with main.  
The CLA and Meta import checks have passed, but the GitHub Actions workflows have not run for the current commit. Could you please re-review the updated PR and approve the workflows when convenient? Please let me know if any further changes are needed. Thanks!

**Comment by ZsyRock** [comment]
Hi May ask any update,  and please let me know if any further changes are needed. Thanks!

**Review by iden-kalemaj — APPROVED** [review]

**Comment by iden-kalemaj** [comment]
Thank you for the changes @ZsyRock, I have accepted the PR and will land it shortly.
